zhaquirks/tuya/ts0601_motion.py

```diff
@@ -66,6 +82,74 @@ class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
     """Tuya local RelativeHumidity cluster."""
 
 
+class TuyaMmwRadarSensitivity(TuyaAttributesCluster, AnalogOutput):
+    """AnalogOutput cluster for sensitivity."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "sensitivity",
+        AnalogOutput.AttributeDefs.min_present_value.id: 1,
+        AnalogOutput.AttributeDefs.max_present_value.id: 9,
+        AnalogOutput.AttributeDefs.resolution.id: 1,
+    }
+
+
+class TuyaMmwRadarMinRange(TuyaAttributesCluster, AnalogOutput):
+    """AnalogOutput cluster for min range."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "min_range",
+        AnalogOutput.AttributeDefs.min_present_value.id: 0,
+        AnalogOutput.AttributeDefs.max_present_value.id: 950,
+        AnalogOutput.AttributeDefs.resolution.id: 10,
+        AnalogOutput.AttributeDefs.engineering_units.id: 118,  # 31: meters
+    }
+
+
+class TuyaMmwRadarMaxRange(TuyaAttributesCluster, AnalogOutput):
+    """AnalogOutput cluster for max range."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "max_range",
+        AnalogOutput.AttributeDefs.min_present_value.id: 10,
+        AnalogOutput.AttributeDefs.max_present_value.id: 950,
+        AnalogOutput.AttributeDefs.resolution.id: 10,
+        AnalogOutput.AttributeDefs.engineering_units.id: 118,  # 31: meters
+    }
+
+
+class TuyaMmwRadarDetectionDelay(TuyaAttributesCluster, AnalogOutput):
+    """AnalogOutput cluster for detection delay."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "detection_delay",
+        AnalogOutput.AttributeDefs.min_present_value.id: 000,
+        AnalogOutput.AttributeDefs.max_present_value.id: 20000,
+        AnalogOutput.AttributeDefs.resolution.id: 100,
+        AnalogOutput.AttributeDefs.engineering_units.id: 159,  # 73: seconds
+    }
+
+
+class TuyaMmwRadarFadingTime(TuyaAttributesCluster, AnalogOutput):
+    """AnalogOutput cluster for fading time."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "fading_time",
+        AnalogOutput.AttributeDefs.min_present_value.id: 2000,
+        AnalogOutput.AttributeDefs.max_present_value.id: 200000,
+        AnalogOutput.AttributeDefs.resolution.id: 1000,
+        AnalogOutput.AttributeDefs.engineering_units.id: 159,  # 73: seconds
+    }
+
+
+class TuyaMmwRadarTargetDistance(TuyaAttributesCluster, AnalogInput):
+    """AnalogInput cluster for target distance."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "target_distance",
+        AnalogOutput.AttributeDefs.engineering_units.id: 118,  # 31: meters
+    }
```

Comment:
I previously thought `target_distance` was supposed to be set from HA as well, but since this is an `AnalogInput` cluster, it means that `target_distance` comes from the sensor and should be exposed (as an output) to ZHA, right?
That's not possible without code changes to ZHA itself yet.

---

I'm currently thinking if we should already move all `AnalogOutput` and `AnalogInput` stuff directly to ZHA (to expose it properly in HA). Otherwise, it probably needs to be done in the future.

For how to set up configuration entities properly in ZHA:
- https://github.com/home-assistant/core/blob/4ee9a6f1301d34c2f287038ebf0c3acdc5f93517/homeassistant/components/zha/number.py#L562-L579
- This would require https://github.com/home-assistant/core/pull/102482 though (to not duplicate all manufactuer/models into ZHA again)

It's not ideal that custom entities can only be created using ZHA at the moment, and it should (hopefully) be possible with quirks directly at some point in the future. Using `AnalogOutput` is a hack though. But it's already done for some other Tuya quirks and I'd really like to get these sensors supported.

I'll need to see if the referenced PR gets merged before the next HA release (sadly doubt that). I'll get back to you/this PR. Sorry for all the waiting. Thanks for all the work on this PR so far!

Reply:
The ideal way forward for this PR would be:
- to remove the `AnalogOutput` clusters (which are used as an input to the device)
- to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
- map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
- assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
- add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
- initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)

That way, the entities are added properly in ZHA. Eventually, I hope that we'll get the functionality to create entities directly via quirks, but this isn't possible at the moment yet.

Reply:
> That would negate the need to write quirks?

No, the Tuya datapoints still have to be mapped to (fake) attributes on a (fake) cluster, as that's the only way we can get that information to ZHA.
However, entity matches in ZHA also need to be added for those attributes.

Reply:
> The ideal way forward for this PR would be:
> 
> * to remove the `AnalogOutput` clusters (which are used as an input to the device)
> * to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
> * map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
> * assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
> * add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
> * initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)
> 
> That way, the entities are added properly in ZHA. Eventually, I hope that we'll get the functionality to create entities directly via quirks, but this isn't possible at the moment yet.

I'll have a look at these items the coming weeks

- [ ] to remove the `AnalogOutput` clusters (which are used as an input to the device)
- [ ] to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
- [ ]  map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
- [ ] assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
- [ ]  add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
- [ ] initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)

Reply:
@TheJulianJES  I did  assign a `quirk_id`, if I copy the file quirk_ids.py in my custom quirks directory, I do not get the import in the  ts0601_motion.py. How do I test this?

Reply:
@TheJulianJES I tried to add the quirk_id, however how to test it?

Would it not be a better plan to merge in the working version of the quirk and than work to improve it with smaller commits. It seems doing everything at once will take a long time, and doing changes over multiple repositories makes it a complex task to do at once where I do not have the time to do test and set it up all at once.

I would suggest to rollback to the working version and that do separate commits to get the other changes in.

Reply:
+1

Reply:
> if I copy the file quirk_ids.py in my custom quirks directory, I do not get the import in the ts0601_motion.py. How do I test this?

You'd need to change the import to something like `from . import xxx`, but the `quirk_id` is just a string, so you don't need to import it from the quirks id file, and can just set it like `quirk_id = "tuya.mmw_radar"`.
In the long run, it's easier to test this with completely replacing the `zhaquirks` package with "a custom one" instead of using custom quirks.
But again, you can just not import it and set `quirk_id = "tuya.mmw_radar"` everywhere in your custom quirk file (or define `TUYA_MMWRADAR = "tuya.mmw_radar"` in your custom quirk file for now).
This can be cleaned up just before merging.

> I tried to add the quirk_id, however how to test it?

It won't do anything on its own, it's just to make it easy to match the quirk to custom entities In ZHA.

> Would it not be a better plan to merge in the working version of the quirk and than work to improve it with smaller commits.

So, the issue with the current version is that we use "hacky" `AnalogInput` and `AnalogOutput` clusters to expose entities in ZHA. If we later decide to properly implement custom entities in ZHA and remove the (fake) `AnalogInput` and `AnalogOutput` clusters from this quirk, the new entities will differ from the old ones.
Entity IDs will change, the old entities will be "no longer provided" but stay in HA until deleted.

Like, I'd be fine to merge the quirk without any configuration entities for now. But I don't think we should include `AnalogInput` / `AnalogOutput` stuff, just to remove it again later (which would be somewhat of a breaking change).

 > It seems doing everything at once will take a long time, and doing changes over multiple repositories makes it a complex task to do at once where I do not have the time to do test and set it up all at once. This would also allow other people to contribute and add more related devices

Of course I understand this though.

> I would suggest to rollback to the working version

Hmm, what changes broke something? Like, what isn't working correctly? Just the quirk ID import?